OpenRA.Mods.Common/UtilityCommands/ImportLegacyMapCommand.cs

```diff
@@ -23,19 +40,329 @@ public bool ValidateArguments(string[] args)
 		}
 
 		[Desc("FILENAME", "Convert a legacy INI/MPR map to the OpenRA format.")]
-		public void Run(ModData modData, string[] args)
+		public virtual void Run(ModData modData, string[] args)
 		{
 			// HACK: The engine code assumes that Game.modData is set.
 			Game.ModData = modData;
 			Game.ModData.MountFiles();
 
-			var rules = Game.ModData.RulesetCache.Load();
-			var map = LegacyMapImporter.Import(args[1], modData.Manifest.Mod.Id, rules, Console.WriteLine);
+			Rules = Game.ModData.RulesetCache.Load();
+
+			var filename = args[1];
+			using (var stream = Game.ModData.ModFiles.Open(filename))
+			{
+				var file = new IniFile(stream);
+				var basic = file.GetSection("Basic");
+				var mapSection = file.GetSection("Map");
+
+				var format = GetMapFormatVersion(basic);
+				ValidateMapFormat(format);
+
+				var tileset = GetTileset(mapSection);
+				Map = new Map(Rules.TileSets[tileset], MapSize, MapSize)
+				{
+					Title = basic.GetValue("Name", Path.GetFileNameWithoutExtension(filename)),
+					Author = "Westwood Studios"
+				};
+
+				Map.Description = ExtractBriefing(file);
+
+				Map.RequiresMod = Game.ModData.Manifest.Mod.Id;
+
+				SetBounds(Map, mapSection);
+
+				ReadPacks(file, filename);
+				ReadTrees(file);
+
+				Map.Videos = LoadVideos(file, "BASIC");
+
+				ReadActors(file);
+
+				LoadSmudges(file, "SMUDGE", MapSize, Map);
+
+				var waypoints = file.GetSection("Waypoints");
+				LoadWaypoints(Map, waypoints, MapSize);
+
+				// Create default player definitions only if there are no players to import
+				MapPlayers = new MapPlayers(Map.Rules, (Players.Count == 0) ? Map.SpawnPoints.Value.Length : 0);
+				foreach (var p in Players)
+					LoadPlayer(file, p);
+				Map.PlayerDefinitions = MapPlayers.ToMiniYaml();
+			}
+
+			Map.FixOpenAreas(Rules);
 
 			var fileName = Path.GetFileNameWithoutExtension(args[1]);
 			var dest = fileName + ".oramap";
-			map.Save(dest);
+			Map.Save(dest);
 			Console.WriteLine(dest + " saved.");
 		}
+
+		/*
+		 * 1=Tiberium Dawn & Sole Survivor
+		 * 2=Red Alert (also with Counterstrike installed)
+		 * 3=Red Alert (with Aftermath installed)
+		 * 4=Tiberian Sun (including Firestorm) & Red Alert 2 (including Yuri's Revenge)
+		 */
+		static int GetMapFormatVersion(IniSection basicSection)
+		{
+			var iniFormat = basicSection.GetValue("NewINIFormat", "0");
+
+			var iniFormatVersion = 0;
+			Exts.TryParseIntegerInvariant(iniFormat, out iniFormatVersion);
+
+			return iniFormatVersion;
+		}
+
+		public abstract void ValidateMapFormat(int format);
+
+		static string ExtractBriefing(IniFile file)
+		{
+			var briefingSection = file.GetSection("Briefing", true);
+			if (briefingSection == null)
+				return string.Empty;
+
+			var briefing = new StringBuilder();
+			foreach (var s in briefingSection)
+				briefing.AppendLine(s.Value);
+
+			return briefing.Replace("\n", " ").ToString();
```

Comment:
It seems that RA used `@` as a new line placeholder. Not sure how C&C handled it since all my maps use mission.ini for their briefing text.